ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,341 @@
+        Xapi_guard.Disk_cache.(Writer.with_cache Tpm (upload session_cache)
+        @@ fun write_tpm -> write_tpm (uuid, time) contents
+    *)
+end = struct
+  let mkdir_safe name perm =
+    Lwt.catch
+      (fun () -> Lwt_unix.mkdir name perm)
+      (function
+        | Unix.(Unix_error (EEXIST, _, _)) -> Lwt.return_unit | e -> raise e
+        )
+
+  let files_in_existing dir =
+    let create_dir = function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          let* () = mkdir_safe dir 0o700 in
+          Lwt.return []
+      | e ->
+          raise e
+    in
+    files_in dir ~otherwise:create_dir
+
+  let persist_to ~filename ~contents =
+    let flags = Unix.[O_WRONLY; O_CREAT; O_SYNC] in
+    let perm = 0o600 in
+    let persist out_chan = Lwt_io.write out_chan contents in
+    Lwt_io.with_file ~flags ~perm ~mode:Output filename persist
+
+  let write_contents ~direct root queue (uuid, now) contents =
+    let __FUN = __FUNCTION__ in
+
+    let tpm_dir = root // Uuidm.(to_string uuid) in
+    (* 1. Record existing requests in cache *)
+    let* outdated_contents = files_in_existing tpm_dir in
+
+    let filename = tpm_dir // (Mtime.to_uint64_ns now |> Int64.to_string) in
+    (* 2. Write new timestamped content to cache, atomically, if needed; and
+       notify the other side, if needed *)
+    let persist () = persist_to ~filename ~contents in
+    let push () =
+      match queue.push (Some (uuid, now)) with
+      | Some () ->
+          Lwt.return_unit
+      | None ->
+          (* Queue is full, change mode to ignore queue *)
+          queue.state <- Disengaged ;
+          Lwt.return_unit
+    in
+    let engage_and_push exn =
+      queue.state <- Engaged ;
+      let* () =
+        Logs_lwt.info (fun m ->
+            m "%s: Error on push. Reason: %s" __FUN (Printexc.to_string exn)
+        )
+      in
+      push ()
+    in
+    let* () =
+      with_lock queue.lock (fun () ->
+          match queue.state with
+          | Direct ->
+              Lwt.try_bind
+                (fun () -> direct (uuid, now) contents)
+                (function
+                  | Ok () -> Lwt.return_unit | Error exn -> engage_and_push exn
+                  )
+                (function exn -> engage_and_push exn)
+          | Engaged ->
+              let* () = persist () in
+              push ()
+          | Disengaged ->
+              persist ()
+      )
+    in
+    (* 4. Delete previous requests from filesystem *)
+    let* _ = Lwt_list.map_p unlink_safe outdated_contents in
+    Lwt.return_unit
+
+  let with_cache ~direct typ queue f =
+    let root = cache_of typ in
+    let* () = mkdir_safe runtime_data 0o700 in
+    let* () = mkdir_safe root 0o700 in
+    f (write_contents ~direct root queue)
+end
+
+module Watcher : sig
+  val watch :
+       direct:(t -> string -> (unit, exn) Lwt_result.t)
+    -> Types.Service.t
+    -> channel
+    -> unit
+    -> unit Lwt.t
+end = struct
+  type action = Content of (t * string)
```

Comment:
Why a variant with a single constructor rather than just `t * string`?

Reply:
This is a leftover from a previous iteration where it was more complex. It doesn't serve any function right now